ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java

```diff
@@ -152,13 +154,18 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, NetworkMetr
         Port port = requestMetadata.requestInfo.getPort();
         String connId = connectionTracker.checkOutConnection(host, port);
         if (connId == null) {
-          if (connectionTracker.mayCreateNewConnection(host, port)) {
+          if (requestMetadata.pendingConnectionId == null && connectionTracker.mayCreateNewConnection(host, port)) {
             connId = selector.connect(new InetSocketAddress(host, port.getPort()), networkConfig.socketSendBufferBytes,
                 networkConfig.socketReceiveBufferBytes, port.getPortType());
             connectionTracker.startTrackingInitiatedConnection(host, port, connId);
-            logger.trace("Initiated a connection to {}:{} ", host, port);
+            requestMetadata.pendingConnectionId = connId;
+            pendingConnectionsToAssociatedRequests.put(connId, requestMetadata);
+            logger.trace("Initiated a connection to host {} port {} ", host, port);
           }
         } else {
+          if (requestMetadata.pendingConnectionId != null) {
+            requestMetadata.pendingConnectionId = null;
```

Comment:
Saw your comment in the `NetworkClient` class confirming that a request need not get the same connection it initiates; want to make sure there is actually no correctness issues.